### PR TITLE
feat(react): enable textarea to limit text input to set amount of words

### LINKED ISSUE #12905: [Feature Request]: word limit for textArea component
### The problem

I would like for the textArea component in the react package to be able to limit the amount of words that user can enter. At the moment it can only limit the amount of characters.

### The solution

Allow textArea component in the react package to count words entered by the user and prevent user from typing more words once the word limit is reached.

### Examples

_No response_

### Application/PAL

_No response_

### Business priority

None

### Available extra resources

_No response_

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)

```[tasklist]
### Tasks
- [ ] https://github.com/carbon-design-system/carbon/issues/13788
- [ ] https://github.com/carbon-design-system/carbon/issues/13789
```

### LINKED ISSUE #13788: TextArea word limit: implementation

### PR DESCRIPTION
**Pull request by matejoslav**
Co-authored-by: SamChinellato <49278203+SamChinellato@users.noreply.github.com>
Co-authored-by: stevenpatrick009 <106097350+stevenpatrick009@users.noreply.github.com>

Closes #13788 

Added functionality to textArea component to limit the number of words the user can type. Also updated the useAnouncer to handle new wording by allowing it to either display "x characters left" or "x words left" based on args passed in.

#### Changelog

**New**

- textArea component has now word counter mode and can now limit the amount of words user types
- added more tests for the textArea component

**Changed**

- useAnouncer can now take the entityName and use that to display the "x <entity name> left"
- added tests for useAnnouncer

#### Testing / Reviewing

- run the storybook for react components
- navigate to the textArea component
- select the "WithWordCounter" textArea type
- type in few words and observe the counter increase
- try to type in more words than the limit and observe that user can not add more words
- try to paste in text with more words than the limit and observe that the text has been cut to match the word limit

### CODE REVIEW

**Comment by matejoslav** [comment]
I have read the DCO document and I hereby sign the DCO.

**Comment by matejoslav** [comment]
recheck

**Comment by sstrubberg** [comment]
@matejoslav thank you so much for the PR. Two things as you're building this out. 

1. Have you considered any error message handling for when a user reaches their limit?
2. Updates to the documentation on the website reflecting this enhancement.

**Comment by matejoslav** [comment]
@sstrubberg Good points.

For point number 2, will make relevant doc updates and push them.

For point number 1, wanted to make the component behave in similar way to having a maximum character limit, at the moment the user cannot exceed the amount of words if they have reached the limit, the announcer message is also displayed showing 0 words left. Not sure if a more visual error handling is needed but happy to add one if you think it would benefit the user experience.

**Comment by matejoslav** [comment]
Just checked the website repo, is it okay to assume that this change would have to be merged first and then we would add the feature description there or would you want them to be done at the same time?

**Comment by sstrubberg** [comment]
> Just checked the website repo, is it okay to assume that this change would have to be merged first and then we would add the feature description there or would you want them to be done at the same time?

correct, merge first, then document. Just didn't want that part to get lost :)

**Comment by tay1orjones** [comment]
@matejoslav Looks like all-contributors needs to be ran again to resolve the conflict with `main`

**Comment by aledavila** [comment]
Bump @matejoslav

**Comment by matejoslav** [comment]
Hello, I am looking into it now 👍

**Comment by matejoslav** [comment]
For some reason I am getting test failure for "Simple date picker › should initialize a calendar when using react.lazy"
I did a clean clone of the carbon repository and getting the same issue.

**Comment by matejoslav** [comment]
Screenshot 2023-01-29 at 17 19 09

**Comment by matejoslav** [comment]
Screenshot 2023-01-29 at 17 30 01

**Comment by tay1orjones** [comment]
@matejoslav It's probably something with your local environment. Try pulling latest from upstream `main` and running `yarn install` and `yarn build`. `yarn test` should pass at that point.

**Comment by matejoslav** [comment]
@tay1orjones the issue is now resolved 👍 was just something strange with my local clone

**Comment by tay1orjones** [comment]
@alisonjoseph @tw15egan I think this is ready for review 👍

File: package.json
Comment (alisonjoseph) [inline package.json]: What was the reason for this change?

File: packages/react/src/components/TextArea/TextArea.js
Comment (alisonjoseph) [inline packages/react/src/components/TextArea/TextArea.js]: Instead of default and words, would it make more sense to be a little more specific and use `character` and `word`

**Review by alisonjoseph — CHANGES_REQUESTED** [review]
This is looking and working great! Couple small suggestions, I'm not sure we need to keep the story for this, leaving it in the playground should be sufficient. I would also update the comment / description for the `enableCounter` prop to just say `Specify whether to display the counter` since it is not longer only used for character count.

**Review by tw15egan — APPROVED** [review]
LGTM once the comments from Alison have been addressed 👍🏻 ✅

**Review by tw15egan — CHANGES_REQUESTED** [review]
It looks like this file was recently converted to Typescript by the Typescript working group helping improve our library. The merge may be a bit messy, it may be best to close this and resubmit starting from the `TextArea.tsx` file

**Comment by matejoslav** [comment]
@alisonjoseph thank you will action your comments

@tw15egan yeah might resubmit, will attempt the merge resolution locally for fun

**Comment by sstrubberg** [comment]
@matejoslav Do you need any help getting this conflicts resolved? Anything you need from us?

**Comment by matejoslav** [comment]
@sstrubberg Hello, just had lack of time but working on it now 👍

**Comment by sstrubberg** [comment]
Hey @matejoslav closing this PR for now. Feel free to re-open once you have your conflicts resolved.

File: package.json
Comment (matejoslav) [reply package.json]: This was because one of the collaborators name was longer than 80 characters and could not be properly included in the commit without increasing the limit for each line. It was the first commit in the branch from December last year

**Comment by matejoslav** [comment]
Just spotted https://github.com/carbon-design-system/carbon/pull/13449

Depending on which PR is merged first, there will be a requirement to update the useEffect to include the behaviour for different counterMode

commenting just in case

File: package.json
Comment (tw15egan) [reply package.json]: It doesn't have to do with the user's name, it's only validating the character count of the commit message. It just needed a shorter commit message.

File: tasks/sync.js
Comment (tw15egan) [inline tasks/sync.js]: Are these changes intentional?

File: package.json
Comment (matejoslav) [reply package.json]: Reverted this change

File: tasks/sync.js
Comment (matejoslav) [reply tasks/sync.js]: ESLint complained locally that these were unused, will check again

Comment (matejoslav) [reply tasks/sync.js]: Screenshot 2023-05-09 at 19 00 19

Comment (tw15egan) [reply tasks/sync.js]: Okay, I was just unsure why this was being pulled in to this PR. Thanks for clarifying.

Comment (alisonjoseph) [reply tasks/sync.js]: Agree, this seems unrelated to the issue/PR

**Comment by matejoslav** [comment]
@tw15egan is the merge-conflict label still required?

**Review by tw15egan — DISMISSED** [review]
This looks great, thanks for sticking with the PR and adding in this functionality! LGTM 👍 ✅

**Comment by tw15egan** [comment]
@matejoslav apologies, but #13449 was merged in before this one and is currently causing a conflict as noted [here](https://github.com/carbon-design-system/carbon/pull/12906#issuecomment-1535459610)

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (francinelucca) [inline packages/react/src/components/TextArea/TextArea.tsx]:
```suggestion
  counterMode?: 'character' | 'word';
```

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (francinelucca) [inline packages/react/src/components/TextArea/TextArea.tsx]: this is really hard to read, could we offload this logic onto a `getInitialTextCount()` function and use if statements instead?

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (francinelucca) [inline packages/react/src/components/TextArea/TextArea.tsx]:
```suggestion
```

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (francinelucca) [inline packages/react/src/components/TextArea/TextArea.tsx]: is a handler still necessary or can we remove this comment?

File: packages/react/src/internal/useAnnouncer.js
Comment (francinelucca) [inline packages/react/src/internal/useAnnouncer.js]: if we set the default of `entityName `to 'characters' here then the if statement below is not necessary and we can always construct the sentence using `${entityName}`
```suggestion
export function useAnnouncer(textCount, maxCount, entityName = 'characters') {
```

**Review by francinelucca — CHANGES_REQUESTED** [review]
This is looking great! just some questions and refactoring suggestions

**Comment by matejoslav** [comment]
@tw15egan yep, resolving the conflict is almost done just got stuck on the tests as the counter changes are using setTimeout which needs accounting for in the tests

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (matejoslav) [reply packages/react/src/components/TextArea/TextArea.tsx]: removed

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (matejoslav) [reply packages/react/src/components/TextArea/TextArea.tsx]: done

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (matejoslav) [reply packages/react/src/components/TextArea/TextArea.tsx]: done

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (matejoslav) [reply packages/react/src/components/TextArea/TextArea.tsx]: done

File: packages/react/src/internal/useAnnouncer.js
Comment (matejoslav) [reply packages/react/src/internal/useAnnouncer.js]: done

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (francinelucca) [inline packages/react/src/components/TextArea/TextArea.tsx]: can we call the getinitialtextcount function here as well?

**Review by francinelucca — CHANGES_REQUESTED** [review]

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (matejoslav) [reply packages/react/src/components/TextArea/TextArea.tsx]: My bad must have missed it

**Comment by matejoslav** [comment]
Could you also please double check there was no regression for https://github.com/carbon-design-system/carbon/pull/13449 I did check but you never know 😃

**Review by francinelucca — CHANGES_REQUESTED** [review]
Confirmed no regression on https://github.com/carbon-design-system/carbon/pull/13449 as a result of these changes. 
However, I do see a weird behavior in that in the current state of the TextArea, setting `enableTextCounter` to false makes the component ignore `maxCount` and doesn't look like that is the case for when counter = 'word' with these new changes, see video below: 

https://github.com/carbon-design-system/carbon/assets/40550942/90f33e28-da71-4d85-9f8d-7c49371eabab

I'd expect counter=word and counter=character to behave the same when it comes to `enableCounter` being on/off and the maxCount being reached, which in production is: the `maxCount` is ignored if `enableCounter` is turned off

**Comment by matejoslav** [comment]
@francinelucca Good spot, I think there might be tests missing for this case so will add them, and this is I think due to the onChange and onKeyDown implementation that does not check if counter is enabled so that should fix it

**Comment by sstrubberg** [comment]
@matejoslav Just wanted to acknowledge you for sticking with this PR and making this contribution with all the back and forth. You rock!

**Comment by matejoslav** [comment]
Got an update, onKeyDown fix is simple as I can basically just add a check for "enableCounter" to be true.

for the onChange function it is a bit tricky. I will post explanation once I test some things out.

**Comment by francinelucca** [comment]
doesn't look like this is ready for re-review @sstrubberg

**Comment by francinelucca** [comment]
Hi @matejoslav 👋🏼, looks like we still have `enableCounter` issue referenced here https://github.com/carbon-design-system/carbon/pull/12906#pullrequestreview-1453799054, will you be able to work on this? We'd love to get this merged in.
TIA

**Comment by matejoslav** [comment]
@francinelucca yes, I have been away for few days but now I am back so will make a change and also post questions about some specific behaviour that I observed

**Comment by matejoslav** [comment]
https://github.com/carbon-design-system/carbon/assets/8973672/7f7cbb31-36d6-407f-ada4-6102b5209179

**Comment by matejoslav** [comment]
@francinelucca please let me know if behaviour in the video is what you would expect, for words when limit is reached and counter is disabled user can continue now typing more words. When counter is enabled again and limit was 5 but user typed in 10 words the next onChange trigger causes the text area's text to be trimmed to the max word count. This is different from the behaviour that was there for characters as the characters that go over the limit stay in the content.

**Comment by matejoslav** [comment]
Once confirmed I will add tests and finalise changes.

**Review by laurenmrice — APPROVED** [review]

**Review by tay1orjones — DISMISSED** [review]
I think with that this should be good to go. @francinelucca can you confirm?

**Comment by tay1orjones** [comment]
@matejoslav looks like there are some conflicts here to address

**Comment by matejoslav** [comment]
@francinelucca @tay1orjones correct, need to add tests, also there appears to be conflict. As far as the behaviour update goes I need to figure out how to handle it, basically the trim was there in case user would paste in text to prevent breaking the limit by pasting in content longer than limit

**Comment by matejoslav** [comment]
Fyi: made some progress with the logic, now it behaves similarly to character, only bug is that the counter does not update immediately when enable counter goes from false to true and there are more words than the limit, it needs an onchange event to register new count so need to check. Also added logic to prevent pasting more words than the limit when enable counter is true

**Comment by matejoslav** [comment]
And it still needs tests :)

**Comment by matejoslav** [comment]
@tay1orjones @francinelucca modified functionality is now pushed and I added missing test, can you please test and let me know. I think the code needs a bit of a refactor so will see what I can do :)

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (francinelucca) [inline packages/react/src/components/TextArea/TextArea.tsx]:
```suggestion
    onKeyDown: (evt) => {
      if (!other.disabled && enableCounter && counterMode === 'word') {
        const key = evt.which;
        if (maxCount && textCount >= maxCount && key === 32) {
          evt.preventDefault();
        }
      }
    },
```

File: packages/react/src/components/TextArea/__tests__/TextArea-test.js
Comment (francinelucca) [inline packages/react/src/components/TextArea/__tests__/TextArea-test.js]:
```suggestion
      rerender(
        <TextArea
          id="input-1"
          labelText="TextArea label"
          enableCounter={true}
          maxCount={5}
          counterMode="word"
        />
```

File: packages/react/src/components/TextArea/__tests__/TextArea-test.js
Comment (francinelucca) [inline packages/react/src/components/TextArea/__tests__/TextArea-test.js]: is this necessary? doesn't seem like it's being used

**Review by francinelucca — CHANGES_REQUESTED** [review]
the enableCounter bug seems to be fixed 🎉. This is almost there, just noticed one detail:
In the video below, I'd expect the counter to correctly show up the number of words/characters when switching between modes


https://github.com/carbon-design-system/carbon/assets/40550942/07b0cabd-e151-494e-97af-bc8a3156f7d0


Also added some cleanup suggestions :)

**Comment by matejoslav** [comment]
Ok let me take a look at the requested changes asap

File: packages/react/src/components/TextArea/__tests__/TextArea-test.js
Comment (matejoslav) [reply packages/react/src/components/TextArea/__tests__/TextArea-test.js]: let me check this, I know I had one working event and one random one

**Comment by matejoslav** [comment]
Interestingly I didnt have conflicts in Readme and contributors file, @francinelucca would you want me to put the contributors at the end of the updated list?

File: packages/react/src/components/TextArea/__tests__/TextArea-test.js
Comment (matejoslav) [reply packages/react/src/components/TextArea/__tests__/TextArea-test.js]: yep you are right its not used

**Comment by matejoslav** [comment]
@francinelucca on the switching of modes and counter update, I think its because it needs onChange to execute at least once to redo the counting, I can try something, potentially manually triggering onChange with the textarea ref to get the counter update or alternatively a hook I guess

File: packages/react/src/components/TextArea/TextArea.tsx
Comment (matejoslav) [inline packages/react/src/components/TextArea/TextArea.tsx:199]: @francinelucca hmm I would expect this to handle the recounting of characters

Comment (matejoslav) [reply packages/react/src/components/TextArea/TextArea.tsx:197]: I think we uncovered a big bug with text area itself, value/defaultValue prop is undefined so when we type the count changes but as soon as you change the mode it tries to calculate initial again and returns 0, that useEffect runs also everytime you type so it gets set to 0 and onChange changes it back to correct value, what was the idea of the useEffect? It was implemented before and I just modified it slightly

Comment (matejoslav) [reply packages/react/src/components/TextArea/TextArea.tsx:199]: Ok the issue is that if value or default value is undefined the change of mode reruns the getInitial count and returns 0, I also discovered that everytime we type the initial count function is reexecuted, that is because there is a re-render of the component (for some reason) and useState reruns the initial value, got a fix I think.

**Comment by francinelucca** [comment]
> Interestingly I didnt have conflicts in Readme and contributors file, @francinelucca would you want me to put the contributors at the end of the updated list?

I don't think order matters that much as long as the file structure is correct, your call!

**Comment by matejoslav** [comment]
Will push the fix tomorrow, finalised it today and also fixed bug in onPaste which allowed people to paste in any content once they had 1 word in text area beforehand

**Comment by matejoslav** [comment]
@francinelucca check please if it now works as expected

**Comment by matejoslav** [comment]
The contributorsrc and Readme seem to have been altered and the order of the people was changed, could you please check and advise what to do?

**Review by francinelucca — APPROVED** [review]
Storybook issue seems to be fixed now, LGTM!

**Comment by francinelucca** [comment]
> The contributorsrc and Readme seem to have been altered and the order of the people was changed, could you please check and advise what to do?

Might be worth to copy/paste the file from main, add your changes and commit to solve conflicts, your call though!

**Review by alisonjoseph — APPROVED** [review]
LGTM!

**Pull request merged**
